modules/phase_field/src/materials/PolycrystalDiffusivity.C

```diff
@@ -0,0 +1,113 @@
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*             Under Contract No. DE-AC07-05ID14517             */
+/*             With the U. S. Department of Energy              */
+/*                                                              */
+/*             See COPYRIGHT for full restrictions              */
+/****************************************************************/
+
+#include "PolycrystalDiffusivity.h"
+#include "libmesh/quadrature.h"
+
+registerMooseObject("PhaseFieldApp", PolycrystalDiffusivity);
+
+InputParameters
+PolycrystalDiffusivity::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription(
+      "Generates a diffusion coefficient to distinguish between the bulk, pore, grain "
+      "boundaries, and surfaces");
+  params.addRequiredCoupledVarWithAutoBuild(
+      "v", "var_name_base", "op_num", "Array of coupled variables");
+  params.addParam<Real>("Dbulk", 1.0, "Diffusion coefficient for volumetric diffusion in solid");
+  params.addParam<Real>(
+      "Dvoid", 1.0, "Diffusion coefficient for diffusion within void/pore/bubble ");
+  params.addParam<Real>("Dsurf", 1.0, "Diffusion coefficient for surface diffusion");
+  params.addParam<Real>("Dgb", 1.0, "Diffusion coefficient for grain boundary diffusion");
+  params.addRequiredCoupledVar("c", "Vacancy phase variable");
+  params.addParam<Real>("surfindex", 1.0, "Surface diffusion index weight");
+  params.addParam<Real>("gbindex", 1.0, "Grain boundary diffusion index weight");
+  params.addParam<Real>("bulkindex", 1.0, "Bulk diffusion index weight");
+  params.addParam<Real>("voidindex", 1.0, "Void diffusion index weight");
+  params.addParam<MaterialPropertyName>(
+      "void_switch", "hb", "Switching Function Materials for the void");
+  params.addParam<MaterialPropertyName>(
+      "solid_switch", "hm", "Switching Function Materials for the solid");
+  params.addParam<MaterialPropertyName>("diffusivity", "D", "The name of the diffusion property");
+  return params;
+}
+
+PolycrystalDiffusivity::PolycrystalDiffusivity(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _c(coupledValue("c")),
+    _c_name(getVar("c", 0)->name()),
+    _diff_name(getParam<MaterialPropertyName>("diffusivity")),
+    _diff(declareProperty<Real>(_diff_name)),
+    _dDdc(declarePropertyDerivative<Real>(_diff_name, _c_name)),
+    _hb(getMaterialProperty<Real>("void_switch")),
+    _hm(getMaterialProperty<Real>("solid_switch")),
+    _dhbdc(getMaterialPropertyDerivative<Real>("void_switch", _c_name)),
+    _dhmdc(getMaterialPropertyDerivative<Real>("solid_switch", _c_name)),
+    _diff_bulk(getParam<Real>("Dbulk")),
+    _diff_void(getParam<Real>("Dvoid")),
+    _diff_surf(getParam<Real>("Dsurf")),
+    _diff_gb(getParam<Real>("Dgb")),
+    _s_index(getParam<Real>("surfindex")),
+    _gb_index(getParam<Real>("gbindex")),
```

Comment:
I don't like the naming of these members. `_index` implies integer valued IMO, like `op_index` below. What those are, are weights. And they should be named accordingly.

Reply:
Changed it.